README
This is Perl module Amon2::Plugin::Web::MobileAgent.

INSTALLATION

Amon2::Plugin::Web::MobileAgent installation is straightforward. If your CPAN shell is set up,
you should just be able to do

% cpan Amon2::Plugin::Web::MobileAgent

Download it, unpack it, then build it as per the usual:

% perl Makefile.PL
% make && make test

Then install it:

% make install

DOCUMENTATION

Amon2::Plugin::Web::MobileAgent documentation is available as in POD. So you can do:

% perldoc Amon2::Plugin::Web::MobileAgent


Changes
Revision history for Perl extension Amon2::Plugin::Web::MobileAgent

{{$NEXT}}

0.11 2013-10-08T23:40:47Z

    - I retired to maintain flavors for Japanese mobile phones.
      We will not create any sites for Japanese mobile phones!
      But, if you want to maintain this module, I can pass the maintainership.
      (tokuhirom)

0.10 2013-08-14T01:53:36Z

    - minil migrate
      (tokuhirom)
    - Fixed setup script around ViewFunctions.
      (tokuhirom)

0.09

    - fixed setup flavor.
      It was broken by Amon2 core's update.

0.08

    - no feature changes.

0.07

    - depend to Encode::JP::Mobile 0.29

0.06

    - fixed testing issue

0.05

    - use MB
    - fixed testing issue

0.04

    - change distribution name
    - fixed testing issue

0.03

    - better supported phones detection
    - app.psgi routes *.psgi
    - better autogenerated routing

0.02

    - MobileJP flavor

0.01    Sat Mar  5 11:22:56 2011
        - original version


t/TestFlavor.pm
use strict;
use warnings;
use utf8;

package t::TestFlavor;
use parent qw(Exporter);
our @EXPORT = qw(test_flavor);
use File::Temp qw/tempdir/;
use App::Prove;
use File::Basename;
use Cwd;
use File::Spec;
use Plack::Util;
use Test::More;

sub test_flavor {
    my ($code, $flavor) = @_;

    local $ENV{PLACK_ENV} = 'development';
    $flavor = Plack::Util::load_class($flavor, 'Amon2::Setup::Flavor');

    my $libpath = File::Spec->rel2abs(File::Spec->catfile(dirname(__FILE__), '..', '..', 'lib'));
    unshift @INC, $libpath;

    my $dir = tempdir(CLEANUP => $ENV{DEBUG} ? 0 : 1);
    my $cwd = Cwd::getcwd();
    chdir($dir);
    note $dir;

    {
        $flavor->new(module => 'My::App')->run;
        $code->($flavor);

        use_ok('t::Util') if -f 't/Util.pm';

        # run prove
        my $app = App::Prove->new();
        $app->process_args('-Ilib', "-I$libpath", <t/*.t>);
        ok($app->run);
    }

    chdir($cwd);
}

1;



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Amon2-MobileJP',
    module_name     => 'Amon2::MobileJP',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Amon2/Plugin/Web/MobileAgent.pm
package Amon2::Plugin::Web::MobileAgent;
use strict;
use warnings;
use 5.008001;
our $VERSION = '0.10';

use HTTP::MobileAgent;

sub init {
    my ($class, $c, $conf) = @_;
    Amon2::Util::add_method(
        $c,
        'mobile_agent',
        sub {
            $_[0]->{mobile_agent} ||= HTTP::MobileAgent->new($_[0]->req->headers);
        }
    );
}

1;
__END__

=encoding utf8

=head1 NAME

Amon2::Plugin::Web::MobileAgent - HTTP::MobileAgent plugin for Amon2

=head1 SYNOPSIS

    package MyApp::Web;
    use parent qw/MyApp Amon2::Web/;
    __PACKAGE__->load_plugins('Web::MobileAgent');
    1;

    # in your controller
    $c->mobile_agent();

=head1 DESCRIPTION

This plugin integrates L<HTTP::MobileAgent> and L<Amon2>.

This module adds C<< $c->mobile_agent() >> method to the context object.
The agent class is generated by C<< $c->req >>.


=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom AAJKLFJEF@ GMAIL COME<gt>

=head1 SEE ALSO

L<HTTP::MobileAgent>, L<Amon2>

=head1 LICENSE

Copyright (C) Tokuhiro Matsuno

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Amon2/Plugin/Web/MobileCharset.pm
package Amon2::Plugin::Web::MobileCharset;
use strict;
use warnings;
use HTTP::MobileAgent::Plugin::Charset;
use Encode::JP::Mobile 0.29;
use Amon2::Util;

sub init {
    my ($class, $c, $conf) = @_;

    Amon2::Util::add_method($c, 'html_content_type' => sub {
        my $ma = shift->mobile_agent;
        my $ct  = $ma->is_docomo ? 'application/xhtml+xml;charset=' : 'text/html;charset=';
           $ct .= $ma->can_display_utf8 ? 'utf-8' : 'Shift_JIS';
           $ct;
    });

    Amon2::Util::add_method($c, 'encoding' => sub {
        shift->mobile_agent->encoding
    });
}

1;
__END__

=encoding utf-8

=head1 NAME

Amon2::Plugin::Web::MobileCharset - Amon2 plugin for Japanese mobile phone's charset

=head1 SYNOPSIS

    package MyApp::Web;
    use Amon2::Web -base;

    __PACKAGE__->load_plugins(
        'Web::MobileAgent',
        'Web::MobileCharset',
    );

=head1 DESCRIPTION

=head1 DEPENDENCIES

This module depend to L<HTTP::MobileAgent::Plugin::Charset>, L<Encode::JP::Mobile>.

And, load L<Amon2::Plugin::MobileAgent> first.

=head1 SEE ALSO

L<HTTP::MobileAgent::Plugin::Charset>, L<Encode::JP::Mobile>



lib/Amon2/Setup/Flavor/MobileJP.pm
use strict;
use warnings;
use utf8;

package Amon2::Setup::Flavor::MobileJP;
use parent qw(Amon2::Setup::Flavor::Large);
use Amon2::Setup::Flavor::Minimum;

sub run {
	my ($self) = @_;
	$self->SUPER::run();

    $self->Amon2::Setup::Flavor::Minimum::create_view(
        tmpl_path => 'tmpl/mobile',
        package => $self->{module} . '::Mobile::View',
        path => "lib/<<PATH>>/Mobile/View.pm",
        view_functions_package => $self->{module} . '::Mobile::ViewFunctions',
    );
    $self->Amon2::Setup::Flavor::Minimum::create_view_functions(
        package => $self->{module} . '::Mobile::ViewFunctions',
        path => "lib/<<PATH>>/Mobile/ViewFunctions.pm",
    );
	$self->write_file("lib/<<PATH>>/Mobile.pm", <<'...');
package <% $module %>::Mobile;
use strict;
use warnings;
use utf8;
use parent qw(<% $module %> Amon2::Web);
use File::Spec;

# dispatcher
use <% $module %>::Mobile::Dispatcher;
sub dispatch {
    return <% $module %>::Mobile::Dispatcher->dispatch($_[0]) or die "response is not generated";
}

use <% $module %>::Mobile::View;
{
    my $view = <% $module %>::Mobile::View->make_instance(__PACKAGE__);
    sub create_view { $view }
}

our $ALLOW_INSECURE_SESSION = 0;

# load plugins
use File::Path qw(mkpath);
use HTTP::Session::State::Cookie;
use HTTP::Session::Store::File;
use File::Spec;
__PACKAGE__->load_plugins(
    'Web::FillInFormLite',
    'Web::NoCache', # do not cache the dynamic content by default
    'Web::CSRFDefender',
	'Web::MobileAgent',
	'Web::MobileCharset',
	'Web::HTTPSession' => {
		state => sub {
			my $c = shift;
			if ($c->mobile_agent->is_docomo && $c->mobile_agent->browser_version < 2.0) {
				if ($ALLOW_INSECURE_SESSION) {
					# insecure.
					require HTTP::Session::State::GUID;
					HTTP::Session::State::GUID->new(
						name => 'amon2_sid',
					);
				} else {
					die "Bad session";
				}
			} else {
				HTTP::Session::State::Cookie->new(
					name => 'amon2_sid',
				);
			}
		},
		store => do {
			my $path = File::Spec->rel2abs(File::Spec->catdir(File::Spec->tmpdir, '<% dist %>-mobile-session'));
			mkpath($path);
			sub {
				my $c = shift;
				HTTP::Session::Store::File->new(
					dir => $path
				)
			}
		},
	},
);

# 全角カタカナを半角カタカナに変換する
use Lingua::JA::Regular::Unicode qw(katakana_z2h);
__PACKAGE__->add_trigger(
    HTML_FILTER => sub {
        my ( $c, $html ) = @_;
        return katakana_z2h($html);
    }
);

__PACKAGE__->add_trigger(
	BEFORE_DISPATCH => sub {
		my ($c) = @_;
		if ($c->is_supported || $c->req->path eq'/non_supported') {
			return; # nop
		} else {
			return $c->redirect('/non_supported');
		}
	},
);

sub is_supported {
	my ($c) = @_;
	my $ma = $c->mobile_agent;
	if ($ma->is_docomo) {
		if ($ma->browser_version < 2.0 && !$ALLOW_INSECURE_SESSION) {
			return 0;
		} elsif ($ma->is_foma) {
			return 1;
		} else {
			return 0; # Mova is not supported
		}
	} elsif ($ma->is_non_mobile) {
		return 0;
	} elsif ($ma->is_softbank) {
		if ($ma->is_3gc) {
			return 1;
		} else {
			return 0;
		}
	} elsif ($ma->is_ezweb) {
		# HDML 端末はもう消失したので、すべてサポートでいいでしょう
		return 1;
	} else {
		return 1;
	}
}

# for your security
__PACKAGE__->add_trigger(
    AFTER_DISPATCH => sub {
        my ( $c, $res ) = @_;
        $res->header( 'X-Content-Type-Options' => 'nosniff' );
    },
);

1;
...

	$self->write_file("lib/<<PATH>>/Mobile/Dispatcher.pm", <<'...');
package <% $module %>::Mobile::Dispatcher;
use strict;
use warnings;
use Router::Simple::Declare;
use Mouse::Util qw(get_code_package);
use Module::Find ();
use String::CamelCase qw(decamelize);

# define roots here.
my $router = router {
	# connect '/' => {controller => 'Root', action => 'index' };
};

my @controllers = Module::Find::useall('<% $module %>::Mobile::C');
{
    no strict 'refs';
    for my $controller (@controllers) {
        my $p0 = $controller;
        $p0 =~ s/^<% $module %>::Mobile::C:://;
        my $p1 = $p0 eq 'Root' ? '' : decamelize($p0) . '/';

        for my $method (sort keys %{"${controller}::"}) {
            next if $method =~ /(?:^_|^BEGIN$|^import$)/;
            my $code = *{"${controller}::${method}"}{CODE};
            next unless $code;
            next if get_code_package($code) ne $controller;
			my $p2 = $method eq 'index' ? '' : $method;
			my $path = "/$p1$p2";
            $router->connect($path => {
                controller => $p0,
                action     => $method,
            });
            print STDERR "map: $path => ${p0}::${method}\n" unless $ENV{HARNESS_ACTIVE};
        }
    }
}

sub dispatch {
    my ($class, $c) = @_;
    my $req = $c->request;
    if (my $p = $router->match($req->env)) {
        my $action = $p->{action};
        $c->{args} = $p;
        "@{[ ref Amon2->context ]}::C::$p->{controller}"->$action($c, $p);
    } else {
        $c->res_404();
    }
}

1;
...

	$self->write_file("lib/<<PATH>>/Mobile/C/Root.pm", <<'...');
package <% $module %>::Mobile::C::Root;
use strict;
use warnings;
use utf8;

sub index {
    my ($class, $c) = @_;
    $c->render('index.tt');
}

sub non_supported {
    my ($class, $c) = @_;
    $c->render('non_supported.tt', {
		insecure_session => $<% $module %>::Mobile::ALLOW_INSECURE_SESSION,
	});
}

1;
...

    $self->write_file('app.psgi', <<'...', {header => $self->psgi_header});
<% $header %>
use <% $module %>::PC;
use Plack::Util;
use Plack::Builder;
use HTTP::MobileAgent;
use Plack::Request;

my $mobile = Plack::Util::load_psgi('mobile.psgi');
my $pc = Plack::Util::load_psgi('pc.psgi');
builder {
    mount '/admin/' => Plack::Util::load_psgi('admin.psgi');
    mount '/m/' => sub {
		my $env = shift;
		if (!HTTP::MobileAgent->new($env->{HTTP_USER_AGENT})->is_non_mobile) {
			return $mobile->($env);
		} else {
			my $req = Plack::Request->new($env);
			my $uri = $req->uri;
			my $path = $uri->path;
			$uri->path('/');
			return [302, ['Location' => $uri->as_string], []];
		}
	};
    mount '/' => sub {
		my $env = shift;
		if (HTTP::MobileAgent->new($env->{HTTP_USER_AGENT})->is_non_mobile) {
			return $pc->($env);
		} else {
			my $req = Plack::Request->new($env);
			my $uri = $req->uri;
			$uri->path('/m/');
			return [302, ['Location' => $uri->as_string], []];
		}
	};
};
...

    $self->write_file('mobile.psgi', <<'...', {header => $self->psgi_header});
<% $header %>
use <% $module %>::Mobile;
use Plack::App::File;
use Plack::Util;
use DBI;

my $basedir = File::Spec->rel2abs(dirname(__FILE__));
my $db_config = <% $module %>->config->{DBI} || die "Missing configuration for DBI";
{
    my $c = <% $module %>->new();
    $c->setup_schema();
}
builder {
    enable 'Plack::Middleware::Static',
        path => qr{^(?:/robots\.txt|/favicon.ico)$},
        root => File::Spec->catdir(dirname(__FILE__), 'static', 'pc');
    enable 'Plack::Middleware::ReverseProxy';

    mount '/static/' => Plack::App::File->new(root => File::Spec->catdir($basedir, 'static', 'pc'));
    mount '/' => <% $module %>::Mobile->to_app();
};
...

	$self->write_file('tmpl/mobile/index.tt', <<'...');
<!doctype html>
<html>
<head>
	<title><% dist %></title>
</head>
<body>
	モバイルサイトの雛形です。
	[% c().mobile_agent().carrier_longname %] でアクセスしています。
</body>
</html>
...

	$self->write_file('tmpl/mobile/non_supported.tt', <<'...');
<!doctype html>
<html>
<head>
	<title><% dist %></title>
</head>
<body>
	<div>非対応端末です。</div>
	<hr />
	対応端末は以下のとおりです。
	<ul>
	<li>ソフトバンク: 3GC</li>
	<li>ドコモ: [% insecure_session ? 'FOMA' : 'ドコモブラウザー2.0以後' %]</li>
	<li>au: WIN</li>
	</ul>
</body>
</html>
...

	$self->write_file('t/01_mobile.t', <<'...');
use strict;
use warnings;
use utf8;
use Test::More;
use Plack::Test;
use Plack::Util;
use t::Util;

my $app = Plack::Util::load_psgi('mobile.psgi');
test_psgi(
	app => $app,
	client => sub {
		my $cb = shift;
		{
			my $req = HTTP::Request->new('GET', '/');
			$req->header('User-Agent' => 'KDDI-CA21 UP.Browser/6.0.6 (GUI) MMP/1.1');
			my $res = $cb->($req);
			is($res->code(), 200) or diag $res->content;
			like($res->decoded_content, qr{EZweb});
			like($res->decoded_content, qr{ﾓﾊﾞｲﾙ}, '半角カタカナフィルタ');
		}
		for my $ua (
			'DoCoMo/1.0/633S/c20',
			'libwww-perl/6.02',
		) {
			subtest $ua =>sub {
				my $req = HTTP::Request->new('GET', '/');
				my $res = $cb->($req);
				is($res->code(), 302) or diag $res->content;
				like($res->header('Location'), qr{/non_supported$});
				my $req2 = HTTP::Request->new('GET' => $res->header('Location'));
				my $res2 = $cb->($req2);
				is($res2->code, 200) or diag $res->as_string;
				like($res2->decoded_content, qr{非対応});
			};
		}
	},
);
done_testing;

...

	$self->write_file('t/05_routing_mobile.t', <<'...');
use strict;
use warnings;
use utf8;
use Test::More;
use Plack::Test;
use Plack::Util;
use t::Util;
use LWP::Protocol::PSGI;
use LWP::UserAgent;

my $app = Plack::Util::load_psgi('app.psgi');
LWP::Protocol::PSGI->register($app);

subtest 'pc browser' => sub {
	my $ua = LWP::UserAgent->new(max_redirect => 0);

	subtest 'cannnot see mobile page' => sub {
		my $res = $ua->get('http://localhost/m/');
		is($res->code, 302) or diag $res->as_string;
		is($res->header('Location'), 'http://localhost/');
	};
	subtest 'can see pc page' => sub {
		my $res = $ua->get('http://localhost/');
		is($res->code, 200) or diag substr($res->as_string, 0, 512);
	};
};

subtest 'mobile browser' => sub {
    my $ua = LWP::UserAgent->new(
        max_redirect => 0,
        agent        => 'KDDI-CA21 UP.Browser/6.0.6 (GUI) MMP/1.1'
    );
	subtest 'cannnot see pc page' => sub {
		my $res = $ua->get('http://localhost/');
		is($res->code, 302) or diag substr($res->as_string, 0, 512);
		is($res->header('Location'), 'http://localhost/m/');
	};
	subtest 'can see mobile page' => sub {
		my $res = $ua->get('http://localhost/m/');
		is($res->code, 200) or diag substr($res->as_string, 0, 512);
	};
};

done_testing;
...

	$self->write_file('t/00_compile_mobile.t', <<'...');
use strict;
use warnings;
use Test::More;

use_ok($_) for qw(
	<% $module %>::Mobile
	<% $module %>::Mobile::Dispatcher
	<% $module %>::Mobile::C::Root
);
done_testing;
...
}

sub create_makefile_pl {
    my ($self, $prereq_pm) = @_;

    $self->SUPER::create_makefile_pl(
        +{
            %{ $prereq_pm || {} },
			'HTTP::MobileAgent' => 0.33,
			'HTTP::MobileAgent::Plugin::Charset' => 0,
			'Encode::JP::Mobile' => '0.29',
			'HTTP::Session::Store::DBI' => '0.02',
			'HTTP::Session' => 0,
			'Amon2::Plugin::Web::HTTPSession' => 0,
			'LWP::Protocol::PSGI' => 0,
			'LWP::UserAgent' => 6,
        },
    );
}

1;
__END__

=for stopwords MobileJP

=encoding utf8

=head1 NAME

Amon2::Setup::Flavor::MobileJP - MobileJP flavor for Amon2

=head1 DESCRIPTION

ガラケーサイト向けのフレーバーです。普通のガラケーサイトで欲しいなあ、とおもうような機能はあらかじめ実装したコードが生成されるようになっています。

このフレーバーは、Large フレーバーをベースにしています。

